### PR TITLE
Add Docker and start glsp-backend with client

### LINKED ISSUE #54: Docker Container during build [8]
Configure the build to produce a minimal docker container

### PR DESCRIPTION
**Pull request by jcebel**
Resolves #54 

Follow up issue: https://github.com/eclipsesource/ecore-glsp/issues/83

Signed-off-by: Jonas Ebel <jonas.ebel@tum.de>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,7 +15,21 @@ Note: to build and run the Ecore GLSP Server, you need Java with version >= 11.
...
- * Follow instructions on /client [Readme file](https://github.com/eclipsesource/ecore-glsp/blob/master/client/README.md)
+## Building and deploying via Docker
+The client repo contains a [Dockerfile](https://github.com/eclipsesource/ecore-glsp/blob/master/client/README.md), that builds the entire client application. The image listens on 0.0.0.0:3000 for incoming requests from a browser.
+
+For installing docker locally please consult [docker's installation description](https://docs.docker.com/install/) for your OS.
+
+**The glsp-server needs to be build locally before you build the image**
+
+**Building**
+`docker build -t <imagename>:<tagname> `
```
Comment (jfaltermeier) [inline README.md]: This misses the dot at the end `docker build -t <imagename>:<tagname> .`

File: client/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,29 @@
+FROM node:10-alpine
+
+RUN mkdir /usr/src/client -p
+
+WORKDIR /usr/src/client
+
+RUN npm install -g yarn
```
Comment (jfaltermeier) [inline client/Dockerfile]: the node:10-alpine image seems to be updated regularly. 
https://hub.docker.com/layers/node/library/node/10-alpine/images/sha256-061cf22fd993245a2eb49a6bd97c32074aa8ea9a35f9a84b7744f2cb2d4c4bb3
was published, as of writing, 10 hours ago and seems to have added yarn out of the box. So installing yarn fails the docker build for me when building without any cached images. 
When removing this line, the docker build works for me. 

Maybe using node:10.0-alpine or node:10.10.0-alpine would work as well?

File: client/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+RUN apk add --update python
+
+RUN apk add --update make
+
+RUN apk add --update g++
+
+RUN apk add --update openjdk11-jre
```
Comment (jfaltermeier) [inline client/Dockerfile]: You could && the run commands in order to create fewer layers. But this may be done alongside issue #83

**Review by jfaltermeier — CHANGES_REQUESTED** [review]
Looks very good. Just one small issue in the documentation and one issue that was caused by a docker base-image image update.

File: client/Dockerfile (hunk 2 above)
Comment (jcebel) [reply client/Dockerfile]: I specified `node:10.18.0-alpine3.10` and removed the explicit yarn statement - this should make the docker build more stable

File: client/Dockerfile (hunk 3 above)
Comment (jcebel) [reply client/Dockerfile]: Done.

**Review by jfaltermeier — DISMISSED** [review]

File: server/ecore-glsp/pom.xml [hunk 4]
```diff
@@ -178,6 +178,24 @@
 					</execution>
 				</executions>
 			</plugin>
+	        <plugin>
+	            <groupId>org.apache.maven.plugins</groupId>
+	            <artifactId>maven-antrun-plugin</artifactId>
+	            <version>1.8</version>
+	            <executions>
+	                <execution>
+	                    <phase>install</phase>
```
Comment (eneufeld) [inline server/ecore-glsp/pom.xml]: wouldn't it be better to bind this to the `verify` phase? This way the package would not be installed into the local mvn repository

File: client/theia-glsp-server/src/node/glsp-server-launcher.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    start(): boolean {
+        if (!this.launchOptions.isRunning) {
+            return this.startServer();
+        }
+        return true;
+    }
+
+    protected startServer(): boolean {
+        if (this.launchOptions.jarPath) {
+            let args = ['-jar', this.launchOptions.jarPath, '--port', `${this.launchOptions.serverPort}`];
```
Comment (eneufeld) [inline client/theia-glsp-server/src/node/glsp-server-launcher.ts]: I'm not sure this `port` passing is working as in https://github.com/eclipsesource/ecore-glsp/blob/8e1b95ee69d3de3ed96c7b2fca32f6ae811827f9/server/ecore-glsp/src/main/java/com/eclipsesource/glsp/ecore/EcoreServerLauncher.java#L29 the port is hardcoded, could you double check?

File: server/ecore-glsp/pom.xml (hunk 4 above)
Comment (jcebel) [reply server/ecore-glsp/pom.xml]: In general I would have assumed, that the developer runs `mvn install ` anyways. Changing the phase for the plugin wouldn't prevent an installation in local repo in that case.

But of course it can be bound differently.

Comment (eneufeld) [reply server/ecore-glsp/pom.xml]: I prefer to run `mvn clean verify` because of the local install otherwise ;-)

Comment (CamilleLetavernier) [reply server/ecore-glsp/pom.xml]: +1 for <= `verify`; installing is only useful if you intend to do partial builds, so it shouldn't be the default phase.

I think `package` would even be the right phase, as `verify` is the integration test phase

File: client/theia-glsp-server/src/node/glsp-server-launcher.ts (hunk 5 above)
Comment (jcebel) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: I've adapted the port retrieval. It is now read from commandline arguments. Please have a look :)

File: server/ecore-glsp/pom.xml (hunk 4 above)
Comment (jcebel) [reply server/ecore-glsp/pom.xml]: Is now part of `package` phase

File: client/theia-glsp-server/src/node/glsp-server-launcher.ts [hunk 6]
```diff
@@ -52,7 +54,11 @@ export class GLSPServerLauncher implements BackendApplicationContribution {
 
     protected startServer(): boolean {
         if (this.launchOptions.jarPath) {
-            let args = ['-jar', this.launchOptions.jarPath, '--port', `${this.launchOptions.serverPort}`];
+            let args = ['-jar', this.launchOptions.jarPath]
```
Comment (eneufeld) [inline client/theia-glsp-server/src/node/glsp-server-launcher.ts]: why this change?

File: server/ecore-glsp/src/main/java/com/eclipsesource/glsp/ecore/EcoreServerLauncher.java [hunk 7]
```diff
@@ -21,12 +21,23 @@
...
 		BasicConfigurator.configure();
+		
 		GLSPServerLauncher launcher = new DefaultGLSPServerLauncher(new EcoreGLSPModule());
+		launcher.start("localhost", port);
+		
+	}
 
-		launcher.start("localhost", 5007);
-
+	private static int searchForPort(String[] args) {
```
Comment (eneufeld) [inline server/ecore-glsp/src/main/java/com/eclipsesource/glsp/ecore/EcoreServerLauncher.java]: Why not call it `getPort` ?

File: client/theia-glsp-server/src/node/glsp-server-launcher.ts (hunk 6 above)
Comment (jcebel) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: In that case we do not have two definitions of the port for the backend (see method getPort at client/theia-ecore/src/node/ecore-glsp-server-contribution.ts)

Comment (eneufeld) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: I don't get it. Aren't you configuring the port in the launchoptions in `CoffeeGlspLaunchOptions` (which should be renamed btw)?
In the previous code iteration you simply accessed the port from the launchoptions and the same is done in https://github.com/eclipsesource/coffee-editor/blob/73db58f116cd75967a663d0a758545ef484031f9/web/coffee-server/src/node/glsp-server-launcher.ts#L55
I think the only necessary change to the previous commit was the change in EcoreServerLauncher.java

Comment (jcebel) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: That is true but in that case I configure the port twice:
Once in the `CoffeeGlspLaunchOptions` .

But the port is also specified as the command line argument that starts theia. (See browser-app/package.json) This is read in the  `ecore-glsp-server-contribution.ts`. 

In my opinion it's best to define this at one place.

Comment (eneufeld) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: @tortmayr do we need the parameter in the package.json?
I would prefer to only keep the value in the launchoptions and get rid of the parameter in the package.json .

Comment (tortmayr) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: No we do not need this parameter.  Keeping the value from the launch options is enough.

I had a quick look at the code and have some additional comments:
- The `theia-glsp-server` module can be merged into the `theia-ecore` module. Splitting this has no real benefit. 
- The jarfiles should not be part of the "src" directory and should be located in a separate directory. (e.g. In the coffee-editor the jars are located a "build" directory)
- In general I think it's not a good idea to check the binary jarfiles in. It would be better to add them to .gitignore and simply always copy the generated jarfile during the build.

Comment (jcebel) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: How do I configure the Ecore-Glsp-Client on which port it should connect, if I remove the `--ECORE_GLSP` from the command line? 

I've understood that the port is read from the Command Args during the creation of the Websocket from Clientside at https://github.com/eclipsesource/ecore-glsp/blob/8e1b95ee69d3de3ed96c7b2fca32f6ae811827f9/client/theia-ecore/src/node/ecore-glsp-server-contribution.ts#L25

Correct Approach would then be to inject the launchConfig at this place?

Comment (eneufeld) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: do it just as was done before by reading the config, see https://github.com/eclipsesource/coffee-editor/blob/73db58f116cd75967a663d0a758545ef484031f9/web/coffee-server/src/node/glsp-server-launcher.ts#L55

Comment (jcebel) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: I'm talking about the `BaseGLSPServerContribution` not the `GLSPServerLauncher`.
What's irritating me is that we have not had a LaunchOptions before.
Maybe it would be interesting to move `GLSPLaunchOptions` into GLSP at some point.
Anyways, I will inject this in both mentioned classes and read the port from there.

Comment (jcebel) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: Please have a look on the code again, if this is now in a state to merge, I will squash the branch one more time :)

Comment (eneufeld) [reply client/theia-glsp-server/src/node/glsp-server-launcher.ts]: @jfaltermeier could you check this?

**Review by eneufeld — DISMISSED** [review]

**Review by eneufeld — APPROVED** [review]
